File: 그래프_dfs/leetcode93(Medium)-Restore IP Addresses.js
// 143ms(5%), 44.7MB(18.35%)
/**
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = function (s) {
  if (Number.isNaN(Number(s))) {
    return []
  }

  const result = []

  const validate = (array) => {
    const last = Number(array[array.length - 1])
    if (last < 0 || last > 255 || array.length > 4) {
      return false
    }
    return true
  }

  const dfs = (index, array) => {
    if (!validate(array)) {
      return
    }

    if (index >= s.length) {
      if (
        array.length !== 4 ||
        array.find((item) => item.length > 1 && item[0] === "0")
      ) {
        return
      }
      result.push(array.join("."))
      return
    }

    const case1 = [...array, s[index]]
    const case2 = [...array]
    case2[case2.length - 1] += s[index]

    dfs(index + 1, case1)
    dfs(index + 1, case2)
  }

  dfs(1, [s[0]])

  return result
}

// 더 깔끔한 방법.. : 69ms(68.7%), 43MB(57.28%)
/**
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = function (s) {
  const result = []

  const isValid = (str) => {
    if (!str || Number(str) > 255) {
      return false
    }
    if (str.length > 1 && str[0] === "0") {
      return false
    }
    return true
  }

  const dfs = (str, array) => {
    if (array.length === 3) {
      if (isValid(str)) {
        result.push([...array, str].join("."))
      }
      return
    }

    for (let i = 1; i < 4; i++) {
      const subStr = str.slice(0, i)
      if (isValid(subStr)) {
        dfs(str.slice(i), [...array, subStr])
      }
    }
  }

  dfs(s, [])

  return result
}

File: 그래프_dfs/leetcode93(Medium)-Restore IP Addresses.py
# 38ms(63.98%), 13.8MB(74.89%)
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        if not s.isdigit():
            return []
        
        result = []

        def dfs(index, array):
            if not 0 <= int(array[-1]) <= 255 or len(array) > 4:
                return
            
            if index >= len(s):
                if len(array) == 4:
                    for item in array:
                        if len(item) > 1 and item[0] == "0":
                            return
                    result.append(".".join(array))
                return
            
            case1 = array[:]
            case1.append(s[index])

            case2 = array[:]
            case2[-1] += s[index]

            dfs(index + 1, case1)
            dfs(index + 1, case2)


        dfs(1, [s[0]])

        return result
    
# 조금 더 나아보이는 방법 : 55ms(21.4%), 13.9MB(74.89%)
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        result = []

        def is_valid(string):
            if not string.isdigit() or int(string) > 255:
                return False
            if len(string) > 1 and string[0] == "0":
                return False
            return True

        def dfs(string, array):
            if len(array) == 3:
                if is_valid(string):
                    result.append(".".join(array + [string]))
                return
            
            for i in range(1, 4):
                substr = string[:i]
                if is_valid(substr):
                    dfs(string[i:], array + [substr])
        
        dfs(s, [])

        return result